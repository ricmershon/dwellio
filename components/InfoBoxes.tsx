import InfoBox from "@/components/InfoBox";

const InfoBoxes = () => (
    <section>
        <div className="container-xl lg:container m-auto">
            <div className="grid grid-cols-1 md:grid-cols-2 gap-4 p-4 rounded-lg">
                <InfoBox
                    headingText="For Renters"
                    backgroundColor="bg-gray-100"
                    buttonInfo={{
                        link: "/properties",
                        styles: "bg-black hover:bg-gray-700",
                        text: "Browse Properties"
                    }}
                >
                    Find your dream rental property. Bookmark properties and contact owners.
                </InfoBox>
                <InfoBox
                    headingText="For Property Owners"
                    backgroundColor="bg-blue-100"
                    buttonInfo={{
                        link: "/properties/add",
                        styles: "bg-blue-500 hover:bg-blue-600",
                        text: "Add Property"
                    }}
                >
                    Find your dream rental property. Bookmark properties and contact owners.
                </InfoBox>
            </div>
        </div>
    </section>
);
 
export default InfoBoxes;