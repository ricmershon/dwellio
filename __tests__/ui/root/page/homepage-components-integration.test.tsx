import React from 'react';
import { render, screen } from '@/__tests__/test-utils';
import { createNextLinkMock } from '@/__tests__/shared-mocks';

// Import the actual components for integration testing
import FeaturedProperties from '@/ui/root/page/featured-properties';
import Hero from '@/ui/root/page/hero';
import InfoBoxes from '@/ui/root/page/info-boxes';

// Mock external dependencies but not the components themselves
jest.mock('next/link', () => createNextLinkMock());

jest.mock('next/image', () => {
    const MockImage = ({ src, alt, className, ...props }: any) => (
        <img 
            src={src} 
            alt={alt} 
            className={className}
            data-testid="property-image"
            {...props}
        />
    );
    MockImage.displayName = 'MockImage';
    return MockImage;
});

// Mock React Suspense
jest.mock('react', () => ({
    ...jest.requireActual('react'),
    Suspense: ({ children }: { children: React.ReactNode }) => (
        <div data-testid="suspense">{children}</div>
    ),
}));

// Mock external utilities and components
jest.mock('@/ui/properties/properties-list', () => {
    const MockPropertiesList = ({ featured, viewportWidth }: { featured: boolean; viewportWidth: number }) => (
        <div 
            data-testid="properties-list" 
            data-featured={featured}
            data-viewport-width={viewportWidth}
        >
            <div data-testid="property-card">Property Card 1</div>
            <div data-testid="property-card">Property Card 2</div>
            <div data-testid="property-card">Property Card 3</div>
        </div>
    );
    MockPropertiesList.displayName = 'MockPropertiesList';
    return MockPropertiesList;
});

// Mock the FeaturedProperties component to avoid async issues in integration tests
jest.mock('@/ui/root/page/featured-properties', () => {
    const MockFeaturedProperties = () => (
        <>
            <h1 className="heading">Featured Properties</h1>
            <div 
                data-testid="properties-list" 
                data-featured="true"
                data-viewport-width="1024"
            >
                <div data-testid="property-card">Property Card 1</div>
                <div data-testid="property-card">Property Card 2</div>
                <div data-testid="property-card">Property Card 3</div>
            </div>
        </>
    );
    MockFeaturedProperties.displayName = 'MockFeaturedProperties';
    return MockFeaturedProperties;
});

jest.mock('@/ui/properties/search/search-form', () => {
    const MockSearchForm = () => (
        <form data-testid="property-search-form">
            <input data-testid="search-input" placeholder="Search properties..." />
            <button type="submit" data-testid="search-button">Search</button>
        </form>
    );
    MockSearchForm.displayName = 'MockSearchForm';
    return MockSearchForm;
});

jest.mock('@/utils/get-viewport-width', () => ({
    getViewportWidth: jest.fn().mockResolvedValue(1024),
}));

// Mock React Icons
jest.mock('react-icons/fa', () => ({
    FaBed: () => <span data-testid="bed-icon">🛏️</span>,
    FaBath: () => <span data-testid="bath-icon">🛁</span>,
    FaRulerCombined: () => <span data-testid="ruler-icon">📏</span>,
    FaMoneyBill: () => <span data-testid="money-icon">💵</span>,
    FaMapMarkerAlt: () => <span data-testid="location-icon">📍</span>,
}));

describe('Homepage Components Integration', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    describe('Component Integration', () => {
        it('should render all homepage components together', () => {
            const HomePageLayout = () => (
                <main>
                    <Hero />
                    <FeaturedProperties />
                    <InfoBoxes />
                </main>
            );

            render(<HomePageLayout />);

            // Verify Hero component elements
            expect(screen.getByText('Find The Perfect Rental')).toBeInTheDocument();
            expect(screen.getByText('Discover the perfect property.')).toBeInTheDocument();
            expect(screen.getByTestId('property-search-form')).toBeInTheDocument();

            // Verify FeaturedProperties component elements
            expect(screen.getByText('Featured Properties')).toBeInTheDocument();
            expect(screen.getByTestId('properties-list')).toBeInTheDocument();

            // Verify InfoBoxes component elements
            expect(screen.getByText('For Renters')).toBeInTheDocument();
            expect(screen.getByText('For Property Owners')).toBeInTheDocument();
            expect(screen.getByText('Browse Properties')).toBeInTheDocument();
            expect(screen.getByText('Add Property')).toBeInTheDocument();
        });

        it('should maintain proper component hierarchy', () => {
            const HomePageLayout = () => (
                <main data-testid="homepage-main">
                    <Hero />
                    <FeaturedProperties />
                    <InfoBoxes />
                </main>
            );

            render(<HomePageLayout />);

            const main = screen.getByTestId('homepage-main');
            const children = Array.from(main.children);

            // Verify order: Hero (section), FeaturedProperties content, InfoBoxes (section)
            expect(children[0].tagName).toBe('SECTION'); // Hero
            expect(children[1]).toHaveTextContent('Featured Properties'); // FeaturedProperties heading
            expect(children[3].tagName).toBe('SECTION'); // InfoBoxes
        });

        it('should handle FeaturedProperties component', () => {
            const HomePageLayout = () => (
                <div>
                    <FeaturedProperties />
                </div>
            );

            render(<HomePageLayout />);

            expect(screen.getByText('Featured Properties')).toBeInTheDocument();
            expect(screen.getByTestId('properties-list')).toBeInTheDocument();

            // Verify viewport width was passed correctly
            const propertiesList = screen.getByTestId('properties-list');
            expect(propertiesList).toHaveAttribute('data-viewport-width', '1024');
            expect(propertiesList).toHaveAttribute('data-featured', 'true');
        });
    });

    describe('Component Communication', () => {
        it('should pass correct props between components', () => {
            render(<FeaturedProperties />);

            const propertiesList = screen.getByTestId('properties-list');
            expect(propertiesList).toHaveAttribute('data-featured', 'true');
            expect(propertiesList).toHaveAttribute('data-viewport-width', '1024');
        });

        it('should render InfoBox components with correct configuration', () => {
            render(<InfoBoxes />);

            // Check that both InfoBox components are rendered with different configurations
            const browseLink = screen.getByRole('link', { name: 'Browse Properties' });
            const addLink = screen.getByRole('link', { name: 'Add Property' });

            expect(browseLink).toHaveAttribute('href', '/properties');
            expect(addLink).toHaveAttribute('href', '/properties/add');

            // Verify different content is rendered
            expect(screen.getByText('Find your dream rental property. Favorite properties and contact owners.')).toBeInTheDocument();
            expect(screen.getByText('Log in and rent your property.')).toBeInTheDocument();
        });
    });

    describe('Responsive Behavior', () => {
        it('should handle different viewport widths in FeaturedProperties', () => {
            // Since we're mocking FeaturedProperties directly, viewport width is fixed
            render(<FeaturedProperties />);
            
            const propertiesList = screen.getByTestId('properties-list');
            expect(propertiesList).toHaveAttribute('data-viewport-width', '1024');
        });

        it('should maintain responsive classes across components', () => {
            render(<Hero />);
            
            const heading = screen.getByText('Find The Perfect Rental');
            expect(heading).toHaveClass('text-4xl', 'sm:text-5xl', 'md:text-6xl');
        });
    });

    describe('Error Handling', () => {
        it('should handle individual component failures gracefully', () => {
            const HomePageLayout = () => (
                <div>
                    <Hero />
                    <InfoBoxes />
                </div>
            );

            render(<HomePageLayout />);

            // Hero and InfoBoxes should render without issues
            expect(screen.getByText('Find The Perfect Rental')).toBeInTheDocument();
            expect(screen.getByText('For Renters')).toBeInTheDocument();
            expect(screen.getByText('For Property Owners')).toBeInTheDocument();
        });
    });

    describe('Performance Integration', () => {
        it('should render all components within reasonable time', () => {
            const startTime = performance.now();
            
            const HomePageLayout = () => (
                <div>
                    <Hero />
                    <FeaturedProperties />
                    <InfoBoxes />
                </div>
            );

            render(<HomePageLayout />);
            
            expect(screen.getByTestId('properties-list')).toBeInTheDocument();
            
            const endTime = performance.now();
            expect(endTime - startTime).toBeLessThan(200);
        });

        it('should handle multiple re-renders efficiently', () => {
            const HomePageLayout = () => (
                <div>
                    <Hero />
                    <FeaturedProperties />
                    <InfoBoxes />
                </div>
            );

            const { rerender } = render(<HomePageLayout />);
            
            expect(screen.getByTestId('properties-list')).toBeInTheDocument();

            // Should handle re-renders without issues
            rerender(<HomePageLayout />);
            
            expect(screen.getByText('Featured Properties')).toBeInTheDocument();
            expect(screen.getByText('Find The Perfect Rental')).toBeInTheDocument();
            expect(screen.getByText('For Renters')).toBeInTheDocument();
        });
    });

    describe('Accessibility Integration', () => {
        it('should maintain proper heading hierarchy across components', () => {
            const HomePageLayout = () => (
                <div>
                    <Hero />
                    <FeaturedProperties />
                    <InfoBoxes />
                </div>
            );

            render(<HomePageLayout />);
            
            expect(screen.getByTestId('properties-list')).toBeInTheDocument();

            const headings = screen.getAllByRole('heading');
            expect(headings.length).toBeGreaterThan(0);
            
            // Verify H1 headings exist
            const h1Headings = headings.filter(h => h.tagName === 'H1');
            expect(h1Headings.length).toBeGreaterThan(0);
        });

        it('should provide accessible navigation across components', () => {
            const HomePageLayout = () => (
                <div>
                    <Hero />
                    <InfoBoxes />
                </div>
            );

            render(<HomePageLayout />);

            const links = screen.getAllByRole('link');
            expect(links.length).toBeGreaterThanOrEqual(2); // InfoBoxes links

            links.forEach(link => {
                expect(link).toHaveAttribute('href');
                expect(link.textContent).toBeTruthy();
            });
        });
    });

    describe('Integration Snapshots', () => {
        it('should match integration snapshot', () => {
            const HomePageLayout = () => (
                <div>
                    <Hero />
                    <FeaturedProperties />
                    <InfoBoxes />
                </div>
            );

            const { container } = render(<HomePageLayout />);
            
            expect(screen.getByTestId('properties-list')).toBeInTheDocument();

            expect(container.firstChild).toMatchSnapshot();
        });
    });
});