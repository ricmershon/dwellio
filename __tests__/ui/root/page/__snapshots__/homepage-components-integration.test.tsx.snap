// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Homepage Components Integration Integration Snapshots should match integration snapshot 1`] = `
<div>
  <section
    class="pt-20 mb-8"
  >
    <div
      class="max-w-7xl mx-auto flex flex-col items-center"
    >
      <div
        class="text-center"
      >
        <h1
          class="text-4xl font-extrabold text-gray-700 sm:text-5xl md:text-6xl"
        >
          Find The Perfect Rental
        </h1>
        <p
          class="mt-4 text-xl text-gray-700"
        >
          Discover the perfect property.
        </p>
      </div>
      <div
        data-testid="suspense"
      >
        <form
          data-testid="property-search-form"
        >
          <input
            data-testid="search-input"
            placeholder="Search properties..."
          />
          <button
            data-testid="search-button"
            type="submit"
          >
            Search
          </button>
        </form>
      </div>
    </div>
  </section>
  <h1
    class="heading"
  >
    Featured Properties
  </h1>
  <div
    data-featured="true"
    data-testid="properties-list"
    data-viewport-width="1024"
  >
    <div
      data-testid="property-card"
    >
      Property Card 1
    </div>
    <div
      data-testid="property-card"
    >
      Property Card 2
    </div>
    <div
      data-testid="property-card"
    >
      Property Card 3
    </div>
  </div>
  <section
    class="mt-8"
  >
    <div
      class="m-auto"
    >
      <div
        class="grid grid-cols-1 md:grid-cols-2 gap-4"
      >
        <div
          class="bg-gray-100 p-6 rounded-lg shadow-md"
        >
          <h1
            class="heading"
          >
            For Renters
          </h1>
          <div
            class="mt-2 mb-4"
          >
            <p
              class="text-sm"
            >
              Find your dream rental property. Favorite properties and contact owners.
            </p>
          </div>
          <a
            class="bg-black hover:bg-gray-700 text-white btn"
            href="/properties"
          >
            Browse Properties
          </a>
        </div>
        <div
          class="bg-blue-100 p-6 rounded-lg shadow-md"
        >
          <h1
            class="heading"
          >
            For Property Owners
          </h1>
          <div
            class="mt-2 mb-4"
          >
            <p
              class="text-sm"
            >
              Log in and rent your property.
            </p>
          </div>
          <a
            class="bg-blue-500 hover:bg-blue-600 text-white btn"
            href="/properties/add"
          >
            Add Property
          </a>
        </div>
      </div>
    </div>
  </section>
</div>
`;
